packages/cli/src/chat/chat-service.ts

```diff
@@ -0,0 +1,298 @@
+
+						const lastNode = getLastNodeExecuted(execution);
+
+						if (lastNode && shouldResumeImmediately(lastNode)) {
+							connection.send('n8n|continue');
+							const data = { action: 'user', chatInput: '', sessionId };
+							await this.resumeExecution(executionId, data, sessionKey);
+							session.waitingNodeName = undefined;
+						} else {
+							session.waitingNodeName = lastNode?.name;
+						}
+					}
+
+					session.isProcessing = false;
+					return;
+				}
+
+				if (execution.status === 'success') {
+					const shouldNotReturnLastNodeResponse =
+						!isPublic || (isPublic && isResponseNodeMode(execution));
+
+					if (shouldNotReturnLastNodeResponse) {
+						closeConnection(connection);
+						session.isProcessing = false;
+						return;
+					}
+
+					const textMessage = prepareMessageFromLastNode(execution);
+
+					connection.send(textMessage, () => {
+						closeConnection(connection);
+					});
+
+					session.isProcessing = false;
+					return;
+				}
+
+				session.isProcessing = false;
+			} catch (error) {
+				if (session) session.isProcessing = false;
+				this.logger.error(
+					`Error sending message to chat in session ${sessionKey}: ${(error as Error).message}`,
+				);
+			}
+		};
+	}
+
+	private incomingMessageHandler(sessionKey: string) {
+		return async (data: RawData) => {
+			try {
+				const session = this.sessions.get(sessionKey);
+
+				if (!session) return;
+
+				const message = data.toString();
+
+				if (message === 'n8n|heartbeat-ack') {
+					session.lastHeartbeat = Date.now();
+					return;
+				}
+
+				const executionId = session.executionId;
+
+				await this.resumeExecution(executionId, this.processIncomingData(data), sessionKey);
+				session.waitingNodeName = undefined;
+			} catch (error) {
+				this.logger.error(
+					`Error processing message from chat in session ${sessionKey}: ${(error as Error).message}`,
+				);
+			}
+		};
+	}
+
+	private async resumeExecution(executionId: string, message: ChatMessage, sessionKey: string) {
+		const execution = await this.getExecution(executionId, sessionKey);
+		if (!execution) return;
+		await this.executionManager.runWorkflow(execution, message);
+	}
+
+	private async getExecution(executionId: string, sessionKey: string) {
+		const execution = await this.executionManager.findExecution(executionId);
+
+		if (!execution || ['error', 'canceled', 'crashed'].includes(execution.status)) {
+			const session = this.sessions.get(sessionKey);
+
+			if (!session) return null;
+
+			session.connection.terminate();
+			clearInterval(session.intervalId);
+			this.sessions.delete(sessionKey);
+			return null;
+		}
+
+		if (execution.status === 'running') return null;
+
+		return execution;
+	}
+
+	private processIncomingData(data: RawData) {
+		const buffer = Array.isArray(data)
+			? Buffer.concat(data.map((chunk) => Buffer.from(chunk)))
+			: Buffer.from(data);
+
+		const message = jsonParse<ChatMessage>(buffer.toString('utf8'));
```

Comment:
We should define a zod schema for the incoming messages and parse them, to ensure we accept only expected messages